src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -811,21 +811,21 @@ trait Infer extends Checkable {
       def onRight = ftpe2 match {
         case OverloadedType(pre, alts)                     => alts forall (alt => isAsSpecific(ftpe1, pre memberType alt))
         case et: ExistentialType                           => et.withTypeVars(isAsSpecific(ftpe1, _))
-        case mt @ MethodType(_, restpe)                    => !mt.isImplicit || isAsSpecific(ftpe1, restpe)
+        case mt @ MethodType(_, restpe)                    => !mt.isImplicit || isAsSpecific(ftpe1, restpe) //
         case NullaryMethodType(res)                        => isAsSpecific(ftpe1, res)
         case PolyType(tparams, NullaryMethodType(restpe))  => isAsSpecific(ftpe1, PolyType(tparams, restpe))
-        case PolyType(tparams, mt @ MethodType(_, restpe)) => !mt.isImplicit || isAsSpecific(ftpe1, PolyType(tparams, restpe))
+        case PolyType(tparams, mt @ MethodType(_, restpe)) => !mt.isImplicit || isAsSpecific(ftpe1, PolyType(tparams, restpe)) //
         case _                                             => isAsSpecificValueType(ftpe1, ftpe2, Nil, Nil)
       }
       ftpe1 match {
         case OverloadedType(pre, alts)                                      => alts exists (alt => isAsSpecific(pre memberType alt, ftpe2))
         case et: ExistentialType                                            => isAsSpecific(et.skolemizeExistential, ftpe2)
         case NullaryMethodType(restpe)                                      => isAsSpecific(restpe, ftpe2)
-        case mt @ MethodType(_, restpe) if mt.isImplicit                    => isAsSpecific(restpe, ftpe2)
+        case mt @ MethodType(_, restpe) if mt.isImplicit                    => isAsSpecific(restpe, ftpe2) //
         case mt @ MethodType(_, _) if bothAreVarargs                        => checkIsApplicable(mt.paramTypes mapConserve repeatedToSingle)
         case mt @ MethodType(params, _) if params.nonEmpty                  => checkIsApplicable(mt.paramTypes)
         case PolyType(tparams, NullaryMethodType(restpe))                   => isAsSpecific(PolyType(tparams, restpe), ftpe2)
-        case PolyType(tparams, mt @ MethodType(_, restpe)) if mt.isImplicit => isAsSpecific(PolyType(tparams, restpe), ftpe2)
+        case PolyType(tparams, mt @ MethodType(_, restpe)) if mt.isImplicit => isAsSpecific(PolyType(tparams, restpe), ftpe2) //
```

Comment:
Can you say a little more about the overloading scenario you would expect to distinguish these two implementations? I've tried the following,
```
bject Overload1 {                                                                                                                    
  def over(i: Int)(implicit d: DummyImplicit): Int = i
  def over(b: Boolean)(implicit d: DummyImplicit): Boolean = b

  (over(23): Int)
  (over(true): Boolean)
}

object Overload2 {
  def over(i: Int)(implicit d: DummyImplicit)(implicit e: DummyImplicit): Int = i
  def over(b: Boolean)(implicit d: DummyImplicit)(implicit e: DummyImplicit): Boolean = b

  (over(23): Int)
  (over(true): Boolean)
}
```
and both appear to compile the same way with both implementations.